src/test/java/lambda/part3/exercise/Mapping.java

```diff
@@ -108,65 +105,92 @@ public void mapping() {
         }
 
-        public <R2> LazyMapHelper<T, R2> map(Function<R, R2> f) {
-            // TODO
-            throw new UnsupportedOperationException();
+        public <R2> LazyMapHelper<T, R2> map(Function<R, R2> mapper) {
+            return new LazyMapHelper<>(list, function.andThen(mapper));
         }
-
     }
 
     private static class LazyFlatMapHelper<T, R> {
+        private final List<T> list;
+        private final Function<T, List<R>> function;
 
         public LazyFlatMapHelper(List<T> list, Function<T, List<R>> function) {
+            this.list = list;
+            this.function = function;
         }
 
         public static <T> LazyFlatMapHelper<T, T> from(List<T> list) {
-            throw new UnsupportedOperationException();
+            return new LazyFlatMapHelper<>(list, Arrays::asList);
         }
 
         public List<R> force() {
-            // TODO
-            throw new UnsupportedOperationException();
+            final List<R> results = new ArrayList<>();
+            for (T element : list) results.addAll(function.apply(element));
+            return results;
         }
 
-        // TODO filter
-        // (T -> boolean) -> (T -> [T])
-        // filter: [T1, T2] -> (T -> boolean) -> [T2]
-        // flatMap": [T1, T2] -> (T -> [T]) -> [T2]
+        public LazyFlatMapHelper<T, R> filter(final Predicate<T> predicate) {
+            return new LazyFlatMapHelper<>(
+                    list,
+                    element -> predicate.test(element)
+                            ? function.apply(element)
+                            : Collections.emptyList()
+            );
+        }
 
-        public <R2> LazyFlatMapHelper<T, R2> map(Function<R, R2> f) {
-            final Function<R, List<R2>> listFunction = rR2TorListR2(f);
-            return flatMap(listFunction);
+        public <R2> LazyFlatMapHelper<T, R2> map(final Function<R, R2> mapper) {
+            return flatMap(rR2TorListR2(mapper));
         }
 
         // (R -> R2) -> (R -> [R2])
-        private <R2> Function<R, List<R2>> rR2TorListR2(Function<R, R2> f) {
-            throw new UnsupportedOperationException();
+        private <R2> Function<R, List<R2>> rR2TorListR2(final Function<R, R2> f) {
+            return element -> Collections.singletonList(f.apply(element));
         }
 
-        // TODO *
-        public <R2> LazyFlatMapHelper<T, R2> flatMap(Function<R, List<R2>> f) {
-            throw new UnsupportedOperationException();
+        public <R2> LazyFlatMapHelper<T, R2> flatMap(final Function<R, List<R2>> mapper) {
+            return new LazyFlatMapHelper<>(
+                    list,
+                    element -> {
+                        final List<R2> results = new ArrayList<>();
+                        function.apply(element)
+                                .forEach(result -> results.addAll(mapper.apply(result)));
+                        return results;
+                    }
+            );
         }
     }
 
-

```

Comment:
Implemented in branch traversable ([pull request](https://github.com/java8-course/lambda/pull/126))